File: android/app/src/main/java/com/nayuta/core2/LspClient.kt
package com.nayuta.core2

import android.annotation.SuppressLint
import android.provider.Settings
import android.util.Log
import com.facebook.react.bridge.*
import com.google.android.gms.tasks.Task
import com.google.android.play.core.integrity.IntegrityManagerFactory
import com.google.android.play.core.integrity.IntegrityServiceException
import com.google.android.play.core.integrity.IntegrityTokenRequest
import com.google.android.play.core.integrity.IntegrityTokenResponse
import com.google.protobuf.ByteString
import com.nayuta.core2.LndGrpc.getLndGrpcString
import lndmobile.Lndmobile
import lspclient.LspClient.IntegrityResult
import java.io.File

class LspClient(private val reactContext: ReactApplicationContext) : ReactContextBaseJavaModule(reactContext) {
    override fun getName(): String {
        return "LspClient"
    }

    @ReactMethod
    fun addListener(eventName: String?) {
        // Keep: Required for RN built in Event Emitter Calls.
    }

    @ReactMethod
    fun removeListeners(count: Double) {
        // Keep: Required for RN built in Event Emitter Calls.
    }

    private fun ByteArray.toHexString() = joinToString("") { "%02x".format(it) }
    private fun String.toHexByteArray(): ByteArray {
        // https://stackoverflow.com/questions/66613717/kotlin-convert-hex-string-to-bytearray
        check(length % 2 == 0) { "odd length" }
        val byteIterator = chunkedSequence(2)
            .map { it.toInt(16).toByte() }
            .iterator()
        return ByteArray(length / 2) { byteIterator.next() }
    }

    @ReactMethod
    fun initialize(
        lspCert: String,
        adminMacaroon: String,
        lspAddr: String,
        lspToken: String,
        promise: Promise
    ) {
        val path = "${reactContext.filesDir}/tls.cert"
        try {
            val lndCert = File(path).readText(Charsets.UTF_8)
            Lndmobile.lcInit(
                lspCert,
                lndCert,
                getLndGrpcString(),
                adminMacaroon,
                lspAddr,
                lspToken
            )
            promise.resolve(null)
        } catch (e: Exception) {
            Log.w(TAG, "connect: ${e.message}")
            promise.reject("connect", e)
        }
    }

    @ReactMethod
    fun ping(nonce: Int, promise: Promise) {
        try {
            val result = Lndmobile.lcPing(nonce)
            promise.resolve(result)
        } catch (e: Exception) {
            promise.reject("ping", e)
        }
    }

    @ReactMethod
    fun getLspVersion(promise: Promise) {
        promise.resolve(Lndmobile.lcGetLspVersion())
    }

    @ReactMethod
    fun getHubLnNodeString(promise: Promise) {
        try {
            promise.resolve(Lndmobile.lcGetHubLnNodeString())
        } catch (e: Exception) {
            Log.w(TAG, "getNodeString: ${e.message}")
            promise.reject("getNodeString", e)
        }
    }

    @ReactMethod
    fun getFeePermyriad(promise: Promise) {
        try {
            promise.resolve(Lndmobile.lcGetLcFeePermyriad().toInt())
        } catch (e: Exception) {
            Log.w(TAG, "getFeePermyriad: ${e.message}")
            promise.reject("getFeePermyriad", e)
        }
    }

    @ReactMethod
    fun receiveMax(promise: Promise) {
        try {
            promise.resolve(Lndmobile.lcReceiveMax().toInt())
        } catch (e: Exception) {
            Log.w(TAG, "receiveMax: ${e.message}")
            promise.reject("receiveMax", e)
        }
    }

    @ReactMethod
    fun paymentFee(reqAmount: Int, promise: Promise) {
        try {
            promise.resolve(Lndmobile.lcPaymentFee(reqAmount.toLong()).toInt())
        } catch (e: Exception) {
            Log.w(TAG, "paymentFee: ${e.message}")
            promise.reject("paymentFee", e)
        }
    }

    @ReactMethod
    fun paymentRegister(reqAmount: Int, memo: String, promise: Promise) {
        try {
            val result = Lndmobile.lcPaymentRegister(reqAmount.toLong(), memo)
            promise.resolve(result)
        } catch (e: Exception) {
            Log.w(TAG, "paymentRegister: ${e.message}")
            promise.reject("paymentRegister", e)
        }
    }

    @ReactMethod
    fun submarineRefundBlock(promise: Promise) {
        promise.resolve(Lndmobile.lcSubmarineRefundBlock())
    }

    @ReactMethod
    fun submarineCreateKeys(promise: Promise) {
        try {
            val result = Lndmobile.lcSubmarineCreateKeys()
            val decoded = lspclient.LspClient.SubmarineCreateKeysResult.parseFrom(result)
            val res = Arguments.createMap()
            res.putString("preimage", decoded.preimage.toByteArray().toHexString())
            res.putString("paymentHash", decoded.paymentHash.toByteArray().toHexString())
            res.putString("repayPrivkey", decoded.repayPrivkey.toByteArray().toHexString())
            res.putString("repayPubkey", decoded.repayPubkey.toByteArray().toHexString())
            promise.resolve(res)
        } catch (e: Exception) {
            Log.w(TAG, "submarineCreateKeys: ${e.message}")
            promise.reject("submarineCreateKeys", e)
        }
    }

    @ReactMethod
    fun submarineRegister(paymentHash: String, repayPubkey: String, promise: Promise) {
        try {
            Log.d(TAG, "submarineRegister: paymentHash=${paymentHash}, repayPubkey=${repayPubkey}")
            val result = Lndmobile.lcSubmarineRegister(paymentHash.toHexByteArray(), repayPubkey.toHexByteArray())
            val decoded = lspclient.LspClient.SubmarineRegisterResult.parseFrom(result)
            val res = Arguments.createMap()
            res.putString("htlcPubkey", decoded.htlcPubkey.toByteArray().toHexString())
            res.putString("script", decoded.script.toByteArray().toHexString())
            res.putString("scriptAddress", decoded.scriptAddress)
            res.putInt("height", decoded.height)
            promise.resolve(res)
        } catch (e: Exception) {
            Log.w(TAG, "submarineRegister: ${e.message}")
            promise.reject("submarineRegister", e)
        }
    }

    @ReactMethod
    fun submarineReceive(paymentHash: String, invoice: String, promise: Promise) {
        try {
            Log.d(TAG, "submarineReceive: paymentHash=${paymentHash}")
            Lndmobile.lcSubmarineReceive(paymentHash.toHexByteArray(), invoice)
            promise.resolve(null)
        } catch (e: Exception) {
            Log.w(TAG, "submarineReceive: ${e.message}")
            promise.reject("submarineReceive", e)
        }
    }

    @ReactMethod
    fun submarineRepayment(repayParams: ReadableArray, repayAddress: String, label: String, promise: Promise) {
        try {
            val requestBuilder = lspclient.LspClient.SubmarineRepayRequest.newBuilder()
            for (idx in 0 until repayParams.size()) {
                val value = repayParams.getMap(idx)
                val repayPrivKey = value.getString("privkey")
                val script = value.getString("script")
                val inTxid = value.getString("txid")
                val inIndex = value.getInt("index")
                val inAmount = value.getInt("amount")
                val repay = lspclient.LspClient.SubmarineRepayData.newBuilder()
                    .setPrivkey(ByteString.copyFrom(repayPrivKey!!.toHexByteArray()))
                    .setScript(ByteString.copyFrom(script!!.toHexByteArray()))
                    .setTxid(inTxid)
                    .setIndex(inIndex)
                    .setAmount(inAmount.toLong())
                    .build()
                requestBuilder.addData(repay)
            }

            val result = Lndmobile.lcSubmarineRepayment(requestBuilder.build().toByteArray(), repayAddress, label)
            Log.d(TAG, "submarineRepayment: txid $result")
            promise.resolve(result)
        } catch (e: Exception) {
            Log.w(TAG, "submarineRepayment: ${e.message}")
            promise.reject("submarineRepayment", e)
        }
    }

    @ReactMethod
    fun submarineReregister(script: String, promise: Promise) {
        try {
            Log.d(TAG, "submarineReregister")
            val result = Lndmobile.lcSubmarineReregister(script.toHexByteArray())
            promise.resolve(result)
        } catch (e: Exception) {
            Log.w(TAG, "submarineReregister: ${e.message}")
            promise.reject("submarineReregister", e)
        }
    }

    @ReactMethod
    fun selfRebalance(promise: Promise) {
        try {
            Lndmobile.lcSelfRebalance()
            promise.resolve(null)
        } catch (e: Exception) {
            Log.w(TAG, "selfRebalance: ${e.message}")
            promise.reject("selfRebalance", e)
        }
    }

    @ReactMethod
    fun queryRoutePayment(invoice: String, feeLimitSat: Int, amtSat: Int) {
        Thread {
            val eventName = "grpc_route_payment"
            try {
                val result = Lndmobile.lcQueryRoutePayment(invoice, feeLimitSat, amtSat.toLong())
                val decoded = lspclient.LspClient.QueryRoutePaymentResult.parseFrom(result)
                val param = Arguments.createMap()
                param.putString("event", "payment")
                param.putString("hash", decoded.paymentHash)
                param.putInt("status", decoded.status)
                param.putInt("failure", decoded.failure)
                LndGrpc.sendEventToJs(reactContext, eventName, param)
            } catch (e: Exception) {
                Log.w(TAG, "selfRebalance: ${e.message}")
                val result = Arguments.createMap()
                result.putString("event", "error")
                result.putString("param", e.message)
                LndGrpc.sendEventToJs(reactContext, eventName, result)
            }
        }.start()
    }

    @ReactMethod
    fun requestOpenChannel(promise: Promise) {
        try {
            Lndmobile.lcRequestOpenChannel()
            promise.resolve(null)
        } catch (e: Exception) {
            Log.w(TAG, "lcRequestOpenChannel: ${e.message}")
            promise.reject("lcRequestOpenChannel", e)
        }
    }

    // get Google Integrity API token
    // https://developer.android.com/google/play/integrity/verdict?hl=ja#request
    @ReactMethod
    fun integrityAppCheck(promise: Promise?) {
        try {
            val id = "" // reserve future use
            val prevResult = Lndmobile.lcIntegrityVerify(id, "")
            if (prevResult != IntegrityResult.INTEGRITYRESULT_NONE.number) {
                Log.d("integrityAppCheck", "prev result = $prevResult")
                promise?.resolve(prevResult == IntegrityResult.INTEGRITYRESULT_OK.number)
                return
            }

            Log.d("integrityAppCheck", "get Token")
            val nonce = Lndmobile.lcIntegrityNonce(id)
            val integrityManager = IntegrityManagerFactory.create(reactContext)
            val integrityTokenResponse: Task<IntegrityTokenResponse> =
                integrityManager.requestIntegrityToken(
                    IntegrityTokenRequest.builder()
                        .setNonce(nonce)
                        .build()
                )
            integrityTokenResponse.addOnSuccessListener {
                Log.d("integrityAppCheck", "get Token: OK")
                try {
                    val verify = Lndmobile.lcIntegrityVerify(id, it.token())
                    Log.d("integrityAppCheck", "new result = $verify")
                    promise?.resolve(verify == IntegrityResult.INTEGRITYRESULT_OK.number)
                } catch (e: Exception) {
                    Log.e("integrityAppCheck", "verify Token NG: ${e.message}")
                    promise?.reject("integrityAppCheck: verify Token NG", e.message)
                }
            }
            integrityTokenResponse.addOnFailureListener {
                Log.e("integrityAppCheck", "get Token: NG: ${it.message}")
                var errCode = ""
                if (it is IntegrityServiceException) {
                    errCode = "(EINTEGRITY=${it.errorCode})"
                    Log.e(
                        "integrityAppCheck",
                        "get Token: NG: message=${it.message}, errorCode=$errCode"
                    )
                }
                promise?.reject(
                    "integrityAppCheck: get Token NG",
                    "Get Token API NG: ${it.message}$errCode)"
                )
            }
        } catch (e: IntegrityServiceException) {
            Log.e("integrityAppCheck", "IntegrityServiceException: message=${e.message}, errorCode=${e.errorCode}, statusCode=${e.statusCode}")
            promise?.reject("integrityAppCheck", e.message)
        } catch (e: Exception) {
            Log.e("integrityAppCheck", "get request Token: NG: ${e.message}")
            promise?.reject("integrityAppCheck", e.message)
        }
    }

    companion object {
        private val TAG = LspClient::class.java.name
    }
}

File: ios/LspClient.swift
//
//  LspClient.swift
//  Core2
//
//  Created by ueno on 2023/04/15.
//

import Foundation
import Lndmobile

@objc(LspClient)
class LspClient : RCTEventEmitter {
  @objc
  override static func requiresMainQueueSetup() -> Bool {
    return false
  }

  @objc
  override func supportedEvents() -> [String]! {
    return ["grpc_route_payment"]
  }

  @objc
  func initialize(
    _ lspCert: String,
    adminMacaroon: String,
    lspAddr: String,
    lspToken: String,
    resolver resolve: @escaping RCTPromiseResolveBlock,
    rejecter reject: @escaping RCTPromiseRejectBlock
  ) {
    print("LspClient.initialize")
    let lndCert = LndReactController.getCertFile()
    LndmobileLcInit(
      lspCert,
      lndCert,
      LndGrpc.getInstance().getLndGrpcString(),
      adminMacaroon,
      lspAddr,
      lspToken
    )
    resolve(nil)
  }

  @objc
  func ping(
    _ nonce: Double,
    resolver resolve: @escaping RCTPromiseResolveBlock,
    rejecter reject: @escaping RCTPromiseRejectBlock
  ) {
    print("LspClient.ping")

    var result: Int32 = 0
    var lndError: NSError?
    LndmobileLcPing(Int32(nonce), &result, &lndError)
    if (lndError == nil) {
      resolve(result)
    } else {
      debugPrint(lndError!)
      reject("LspClient.ping", lndError?.localizedDescription, lndError)
    }
  }

  @objc
  func getLspVersion(
    _ resolve: @escaping RCTPromiseResolveBlock,
    rejecter reject: @escaping RCTPromiseRejectBlock
  ) {
    print("LspClient.getLspVersion")

    resolve(LndmobileLcGetLspVersion())
  }

  @objc
  func getHubLnNodeString(
    _ resolve: @escaping RCTPromiseResolveBlock,
    rejecter reject: @escaping RCTPromiseRejectBlock
  ) {
    print("LspClient.getHubLnNodeString")

    var lndError: NSError?
    let result = LndmobileLcGetHubLnNodeString(&lndError)
    if (lndError == nil) {
      resolve(result)
    } else {
      reject("LspClient.getHubLnNodeString", lndError?.localizedDescription, lndError);
    }
  }

  @objc
  func getFeePermyriad(
    _ resolve: @escaping RCTPromiseResolveBlock,
    rejecter reject: @escaping RCTPromiseRejectBlock
  ) {
    print("LspClient.getFeePermyriad")

    var result: Int64 = 0
    var lndError: NSError?
    LndmobileLcGetLcFeePermyriad(&result, &lndError)
    if (lndError == nil) {
      resolve(result)
    } else {
      reject("LspClient.getFeePermyriad", lndError?.localizedDescription, lndError);
    }
  }

  @objc
  func receiveMax(
    _ resolve: @escaping RCTPromiseResolveBlock,
    rejecter reject: @escaping RCTPromiseRejectBlock
  ) {
    print("LspClient.receiveMax")

    var result: Int64 = 0
    var lndError: NSError?
    LndmobileLcReceiveMax(&result, &lndError)
    if (lndError == nil) {
      resolve(result)
    } else {
      reject("LspClient.receiveMax", lndError?.localizedDescription, lndError);
    }
  }

  @objc
  func paymentFee(
    _ reqAmount: Double,
    resolver resolve: @escaping RCTPromiseResolveBlock,
    rejecter reject: @escaping RCTPromiseRejectBlock
  ) {
    print("LspClient.paymentFee")

    var result: Int64 = 0
    var lndError: NSError?
    LndmobileLcPaymentFee(Int64(reqAmount), &result, &lndError)
    if (lndError == nil) {
      resolve(result)
    } else {
      reject("LspClient.paymentFee", lndError?.localizedDescription, lndError);
    }
  }

  @objc
  func paymentRegister(
    _ reqAmount: Double,
    memo: String,
    resolver resolve: @escaping RCTPromiseResolveBlock,
    rejecter reject: @escaping RCTPromiseRejectBlock
  ) {
    print("LspClient.paymentRegister")

    var lndError: NSError?
    let result = LndmobileLcPaymentRegister(Int64(reqAmount), memo, &lndError)
    if (lndError == nil) {
      resolve(result)
    } else {
      print("paymentRegister error: \(lndError as Any)")
      reject("LspClient.paymentRegister error", lndError?.localizedDescription, lndError)
    }
  }

  @objc
  func submarineRefundBlock(
    _ resolve: @escaping RCTPromiseResolveBlock,
    rejecter reject: @escaping RCTPromiseRejectBlock
  ) {
    print("LspClient.submarineRefundBlock")

    let result = LndmobileLcSubmarineRefundBlock()
    resolve(result)
  }

  @objc
  func submarineCreateKeys(
    _ resolve: @escaping RCTPromiseResolveBlock,
    rejecter reject: @escaping RCTPromiseRejectBlock
  ) {
    print("LspClient.submarineCreateKeys")

    var lndError: NSError?
    let result = LndmobileLcSubmarineCreateKeys(&lndError)
    if (lndError != nil) {
      print("submarineCreateKeys error: \(lndError as Any)")
      reject("LspClient.submarineCreateKeys error", lndError?.localizedDescription, lndError)
      return
    }
    do {
      let decoded = try Lspclient_SubmarineCreateKeysResult.init(serializedData: result! as Data)
      let res: [String: Any] = [
        "preimage": dataToHex(decoded.preimage),
        "paymentHash": dataToHex(decoded.paymentHash),
        "repayPrivkey": dataToHex(decoded.repayPrivkey),
        "repayPubkey": dataToHex(decoded.repayPubkey)
      ]
      resolve(res)
    } catch {
      print("submarineCreateKeys.Lspclient_SubmarineCreateKeysResult error: \(lndError as Any)")
      reject("LspClient.submarineCreateKeys.Lspclient_SubmarineCreateKeysResult error", lndError?.localizedDescription, error)
    }
  }

  @objc
  func submarineRegister(
    _ paymentHash: String,
    repayPubkey: String,
    resolver resolve: @escaping RCTPromiseResolveBlock,
    rejecter reject: @escaping RCTPromiseRejectBlock
  ) {
    print("LspClient.submarineRegister");

    let paymentHashData = hexToData(paymentHash)
    let repayPubkeyData = hexToData(repayPubkey)
    if (paymentHashData == nil || repayPubkeyData == nil) {
      reject("LspClient.submarineRegister: nil data", "LspClient.submarineRegister: nil data", nil)
      return
    }
    var lndError: NSError?
    let result = LndmobileLcSubmarineRegister(paymentHashData, repayPubkeyData, &lndError)
    if (lndError != nil) {
      reject("LspClient.submarineRegister", lndError?.localizedDescription, lndError)
      return
    }
    do {
      let decoded = try Lspclient_SubmarineRegisterResult.init(serializedData: result! as Data)
      let res: [String: Any] = [
        "htlcPubkey": dataToHex(decoded.htlcPubkey),
        "script": dataToHex(decoded.script),
        "scriptAddress": decoded.scriptAddress,
        "height": decoded.height
      ]
      resolve(res)
    } catch {
      print("submarineCreateKeys.Lspclient_SubmarineRegisterResult error: \(error as Any)")
      reject("LspClient.submarineCreateKeys.Lspclient_SubmarineRegisterResult error", error.localizedDescription, error)
    }
  }

  @objc
  func submarineReceive(
    _ paymentHash: String,
    invoice: String,
    resolver resolve: @escaping RCTPromiseResolveBlock,
    rejecter reject: @escaping RCTPromiseRejectBlock
  ) {
    print("LspClient.submarineReceive");

    var lndError: NSError?
    let result = LndmobileLcSubmarineReceive(hexToData(paymentHash), invoice, &lndError)
    if (lndError == nil) {
      resolve(result)
    } else {
      print("submarineReceive error: \(lndError as Any)")
      reject("LspClient.submarineReceive error", lndError?.localizedDescription, lndError)
    }
  }

  @objc
  func submarineRepayment(
    _ repayParams: [Dictionary<String, Any>],
    repayAddress: String,
    label: String,
    resolver resolve: @escaping RCTPromiseResolveBlock,
    rejecter reject: @escaping RCTPromiseRejectBlock
  ) {
    print("LspClient.submarineRepayment");

    do {
      var req = Lspclient_SubmarineRepayRequest()
      for lp in 0 ..< repayParams.count {
        var repay = Lspclient_SubmarineRepayData()
        repay.privkey = hexToData(repayParams[lp]["privkey"] as! String)!
        repay.script = hexToData(repayParams[lp]["script"] as! String)!
        repay.txid = repayParams[lp]["txid"] as! String
        repay.index = repayParams[lp]["index"] as! Int32
        repay.amount = repayParams[lp]["amount"] as! Int64
        req.data.append(repay)
      }
      var lndError: NSError?
      let result = try LndmobileLcSubmarineRepayment(req.serializedData(), repayAddress, label, &lndError)
      if (lndError == nil) {
        resolve(result)
      } else {
        print("submarineRepayment error: \(lndError as Any)")
        reject("LspClient.submarineRepayment error", lndError?.localizedDescription, lndError)
      }
    } catch {
      print("submarineRepayment throw error: \(error)")
      reject("LspClient.submarineRepayment throw error", error.localizedDescription, error)
    }
  }

  @objc
  func submarineReregister(
    _ script: String,
    resolver resolve: @escaping RCTPromiseResolveBlock,
    rejecter reject: @escaping RCTPromiseRejectBlock
  ) {
    print("LspClient.submarineReregister");

    var lndError: NSError?
    let result = LndmobileLcSubmarineReregister(hexToData(script), &lndError)
    if (lndError == nil) {
      resolve(result)
    } else {
      print("submarineReregister error: \(lndError as Any)")
      reject("LspClient.submarineReregister error", lndError?.localizedDescription, lndError)
    }
  }

  @objc
  func selfRebalance(
    _ resolve: @escaping RCTPromiseResolveBlock,
    rejecter reject: @escaping RCTPromiseRejectBlock
  ) {
    print("LspClient.selfRebalance");

    var lndError: NSError?
    let result = LndmobileLcSelfRebalance(&lndError)
    if (lndError == nil) {
      resolve(result)
    } else {
      print("selfRebalance error: \(lndError as Any)")
      reject("LspClient.selfRebalance error", lndError?.localizedDescription, lndError)
    }
  }

  @objc
  func queryRoutePayment(
    _ invoice: String,
    feeLimitSat: Double,
    amtSat: Double
  ) {
    print("LspClient.queryRoutePayment");
    let eventName = "grpc_route_payment"

    func errorHandler(_ error: String?) {
      let result: [String: Any] = [
        "event": "error",
        "param": error ?? "unknown error"
      ]
      sendEvent(withName: eventName, body: result)
    }

    var lndError: NSError?
    let result = LndmobileLcQueryRoutePayment(invoice, Int32(feeLimitSat), Int64(amtSat), &lndError)
    if (lndError != nil) {
      let err = lndError?.localizedDescription ?? "unknown error"
      print("submarineCreateKeys.queryRoutePayment error: \(err)")
      errorHandler(err)
      return
    }
    do {
      let decoded = try Lspclient_QueryRoutePaymentResult.init(serializedData: result! as Data)
      let result: [String: Any] = [
        "event": "payment",
        "hash": decoded.paymentHash,
        "status": decoded.status,
        "failure": decoded.failure
      ]
      sendEvent(withName: eventName, body: result)
    } catch {
      print("submarineCreateKeys.queryRoutePayment throw error: \(error)")
      errorHandler("\(error)")
    }
  }

  @objc
  func requestOpenChannel(
    _ resolve: @escaping RCTPromiseResolveBlock,
    rejecter reject: @escaping RCTPromiseRejectBlock
  ) {
    print("LspClient.requestOpenChannel");

    var lndError: NSError?
    let result = LndmobileLcRequestOpenChannel(&lndError)
    if (lndError == nil) {
      resolve(result)
    } else {
      print("requestOpenChannel error: \(lndError as Any)")
      reject("LspClient.requestOpenChannel error", lndError?.localizedDescription, lndError)
    }
  }

  @objc
  func integrityAppCheck(
    _ resolve: @escaping RCTPromiseResolveBlock,
    rejecter reject: @escaping RCTPromiseRejectBlock
  ) {
    var lndError: NSError?
    let id = "iOS"

    var prevResult: Int32 = 0
    LndmobileLcIntegrityVerify(id, "", &prevResult, &lndError)
    if (lndError != nil) {
      print("integrityAppCheck verify error: \(lndError as Any)")
      reject("LspClient.integrityAppCheck verify error", lndError?.localizedDescription, lndError)
      return
    }
    if (prevResult != Lspclient_IntegrityResult.none.rawValue) {
      print("LspClient.integrityAppCheck: prev result=\(prevResult)")
      resolve(prevResult == Lspclient_IntegrityResult.ok.rawValue)
      return
    }
    let nonce = LndmobileLcIntegrityNonce(id, &lndError)
    if (lndError != nil) {
      print("integrityAppCheck nonce error: \(lndError as Any)")
      reject("LspClient.integrityAppCheck nonce error", lndError?.localizedDescription, lndError)
      return
    }
    print("LspClient.integrityAppCheck: nonce=\(nonce)")
    var verify: Int32 = 0
    LndmobileLcIntegrityVerify(id, nonce, &verify, &lndError)
    print("LspClient.integrityAppCheck: verify=\(verify)")
    resolve(verify == Lspclient_IntegrityResult.ok.rawValue)
  }
}
